src/Adapter/Carrier/Repository/CarrierRangeRepository.php

```diff
@@ -0,0 +1,249 @@
+use IteratorAggregate;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\Exception\CarrierConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\Exception\CarrierException;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierId;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierRange;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierRangePrice;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+
+/**
+ * Provides access to carrier range data source
+ */
+class CarrierRangeRepository
+{
+    public function __construct(
+        protected readonly Connection $connection,
+        protected readonly string $dbPrefix,
+        protected readonly CarrierRepository $carrierRepository,
+    ) {
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     *
+     * @return CarrierRange[]
+     *
+     * @throws CarrierException
+     * @throws \Doctrine\DBAL\Exception
+     * @throws \PrestaShop\PrestaShop\Core\Domain\AttributeGroup\Attribute\Exception\AttributeNotFoundException
+     * @throws CarrierConstraintException
+     * @throws \PrestaShop\PrestaShop\Core\Domain\Zone\Exception\ZoneException
+     * @throws \PrestaShop\PrestaShop\Core\Exception\CoreException
+     */
+    public function get(CarrierId $carrierId, ShopConstraint $shopConstraint): array
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $carrier = $this->carrierRepository->get($carrierId);
+
+        // Then, create query to retrieve carrier ranges
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('cd.id_zone, cr.delimiter1 AS range_from, cr.delimiter2 AS range_to, cd.price AS range_price')
+            ->from($this->dbPrefix . 'delivery', 'cd')
+            ->andWhere('cd.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue());
+
+        // Apply shipping method (weight or price) and shop constraint, then execute the query
+        $this->applyRangeTypeForQuery($qb, $carrier);
+        $this->applyShopConstraint($qb, $shopConstraint);
+        $results = $qb->executeQuery()->fetchAllAssociative();
+
+        // Format results with CarrierRange and CarrierRangePrice objets
+        return self::formatRangesFromData($results);
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     * @param array $ranges
+     * @param ShopConstraint $shopConstraint
+     *
+     * @return CarrierId
+     */
+    public function set(CarrierId $carrierId, array $ranges, ShopConstraint $shopConstraint): CarrierId
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $carrier = $this->carrierRepository->get($carrierId);
+        $rangeTable = $this->getRangeTypeTable($carrier);
+
+        // Use transaction to ensure data consistency
+        $this->connection->beginTransaction();
```

Comment:
Nice idea, but I'm not sure transaction actually work 🤔
Oh no they should work with a Doctrine connection